### ISSUE TITLE
Revisit public API analyzer design

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn-analyzers#5787]

**Comment by jmarolf** [comment]
@mikadumont my opinion is that this is a general feature that all .NET developers could benefit from. Unclear at this point if this would be a .NET SDK only feature or if we should have some additional UI in visual studio for it but I would like to start the conversation.

**Comment by mavasani** [comment]
Agree with @jmarolf. A prescribed way to track public API surface of libraries should be something that all .NET developers need to do and a design with collaboration with the.NET libraries/runtime team would be beneficial. It should also likely ship inside the .NET SDK. We may also want to consider if there is benefit allowing including some notion of versioning in the shipped API file so shipped APIs are grouped by released versions, and we can have different  viewers to show views with different API groupings based on user scenario.

cc @terrajobst @jeffhandley

**Comment by mavasani** [comment]
Also tagging @333fred for his views, given his experience driving Roslyn public API reviews.

**Comment by ericstj** [comment]
cc @safern @terrajobst We shipped a solution in 6.0 that addresses this scenario.  https://devblogs.microsoft.com/dotnet/package-validation/

It leverages the previously shipped assemblies and roslyn object model as the source for API.

Because we wished to also compare across active TargetFrameworks, and we wanted the flexibility to run binary-to-binary tool, we opted to make it post-package validation.  @safern and @terrajobst have a number of ideas for improving it.

**Comment by jmarolf** [comment]
Got some nice people at nuget to get me a list of all public repos that depend on this package.

- .NET Repos
  - https://github.com/dotnet/format
  - https://github.com/dotnet/msbuild
  - https://github.com/dotnet/roslyn-analyzers
  - https://github.com/dotnet/Silk.NET
  - https://github.com/dotnet/templating
  - https://github.com/dotnet/winforms

- Microsoft Repos
  - https://github.com/microsoft/ApplicationInsights-dotnet
  - https://github.com/microsoft/OpenAPI.NET.OData
  - https://github.com/microsoft/vs-mef
  - https://github.com/microsoft/vs-streamjsonrpc
  - https://github.com/microsoft/vs-threading
  - https://github.com/microsoft/vstest
  - https://github.com/NuGet/NuGet.Client
  - https://github.com/OfficeDev/Open-XML-SDK

- Large Projects
  - https://github.com/DotNetAnalyzers/StyleCopAnalyzers
  - https://github.com/OData/AspNetCoreOData
  - https://github.com/open-telemetry/opentelemetry-dotnet

- Other Projects
  - https://github.com/AArnott/Nerdbank.Streams
  - https://github.com/adamralph/bullseye
  - https://github.com/adamralph/liteguard
  - https://github.com/adamralph/simple-exec
  - https://github.com/alpacahq/alpaca-trade-api-csharp
  - https://github.com/ashmind/mirrorsharp
  - https://github.com/atifaziz/NCrontab
  - https://github.com/ChilliCream/hotchocolate
  - https://github.com/docopt/docopt.net
  - https://github.com/drewnoakes/metadata-extractor-dotnet
  - https://github.com/GuOrg/Gu.Wpf.DataGrid2D
  - https://github.com/justeat/httpclient-interception
  - https://github.com/maxhauser/semver
  - https://github.com/natemcmaster/CommandLineUtils
  - https://github.com/natemcmaster/DotNetCorePlugins
  - https://github.com/natemcmaster/LettuceEncrypt
  - https://github.com/neuecc/MessagePack-CSharp
  - https://github.com/npgsql/npgsql

**Comment by jmarolf** [comment]
As @ericstj I think we should try and find a path forward to use package validation for API enforcement. If there are gaps here and Roslyn can't adopt it for some reason, we should go make the changes in the [code](https://github.com/dotnet/sdk/tree/main/src/Compatibility/Microsoft.DotNet.ApiCompatibility) to handle these situations

A few gaps that we should ask ourselves if we care about:
- package validation only runs on `dotnet pack` so you would get no warning that you broke things until you ran CI or manually tried to run `pack` again. Today PublicApiAnalyzers can give you this information in the IDE
- There is apparently a [bespoke system](https://docs.microsoft.com/en-us/dotnet/fundamentals/package-validation/overview#suppress-compatibility-errors) that was created for suppressing packages validation errors. Need to see what the tooling is like for this is VS. Unfortunate that we would have `.editorconfig`, `ruleset` and now `CompatibilitySuppressions.xml` as options for configuring suppressing warnings.

**Comment by safern** [comment]
> A few gaps that we should ask ourselves if we care about:

I completely agree about these gaps and this is something that we are planning to invest on .NET 7. The reason why we went with dotnet pack first was because it felt easier to adopt by a regular library/customer and we tried to think out of the box rather than just what our repositories need. Also the package validation we run after `dotnet pack` not only checks `API Compatibility` it also checks that the package is well formed and other validations that are helpful for customers that are producing a package.

However, that said, we have planned adding an MSBuild entry point to be able to run only API Compat as part of the build without the need to run `dotnet pack` and controlled via an MSBuild switch. This would permit using a `.cs` file as the contract that the implementation needs to meet, or a binary. 

The reason we didn't go with an analyzer to begin with was already described by @ericstj because we wanted binary-to-binary compatibility, however I do see this code being able to be expanded to take either a binary previously compiled or a `.cs`  file as a contract and then running as an analyzer during a design time compilation. However for this we would need to optimize in performance perhaps to make sure we run fast enough and use better caching mechanisms, but all is doable given that we are already built on top of Roslyn. We could even support different TFMs by specifying the binaries or `.cs` files via analyzer configurations that are specific to a TFM (I believe that's doable but not sure?).

To add to this, we still have some gaps we want to fill on rules, but it would be great if more teams join the effort so that we can achieve these goals 😄.

**Comment by Evangelink** [comment]
> but the txt file format is not at all useful for driving reviews.

This request https://github.com/dotnet/roslyn-analyzers/issues/3901 was suggesting to have a C# syntax (could potentially also be vb for those who prefer). Maybe we could get some inspiration from F# (or other languages) signature files.

**Comment by jmarolf** [comment]
> This request #3901 was suggesting to have a C# syntax

@333fred and I talked in person about this (weird I know!) and I think the general consensus from everyone is that we want a C#-like syntax. 

By basic sketch for this is:
- You have a source generator run and compute the public api for a compilation (same rules that we use for the Public API Analyzer today)
- The source generator then creates a file (let's call it `Public.Apis.csi` for now) an that file gets checked in.
- On CI you have some MSBuild property that gets passed to the analyzer/source generator to fail the build if the public API doesn't match the source (so if someone changes the `Public.Apis.csi` file and forgets to check it in we'll fail the build)

The *.csi files (or whatever we call them) would have same format as meta-as-source, meaning that instead of a body they would have `;` character. This would mean that changes in comments and attributes would now be visible.


Here is a short example of the output on the string type (I elided all the members but two for brevity):
```csharp
#nullable enable

using System.Buffers;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using System.Reflection;
using System.Runtime.CompilerServices;
using System.Text;

namespace System
{
    //
    // Summary:
    //     Represents text as a sequence of UTF-16 code units.
    [DefaultMember("Chars")]
    public sealed class String : IEnumerable<char>, IEnumerable, ICloneable, IComparable, IComparable<String?>, IConvertible, IEquatable<String?>
    {
        //
        // Summary:
        //     Represents the empty string. This field is read-only.
        public static readonly String Empty;

        //
        // Summary:
        //     Initializes a new instance of the System.String class to the value indicated
        //     by a specified pointer to an array of Unicode characters.
        //
        // Parameters:
        //   value:
        //     A pointer to a null-terminated array of Unicode characters.
        //
        // Exceptions:
        //   T:System.ArgumentOutOfRangeException:
        //     The current process does not have read access to all the addressed characters.
        //
        //   T:System.ArgumentException:
        //     value specifies an array that contains an invalid Unicode character, or value
        //     specifies an address less than 64000.
        [CLSCompliant(false)]
        public String(char* value);
        //
        // Summary:
        //     Initializes a new instance of the System.String class to the Unicode characters
        //     indicated in the specified character array.
        //
        // Parameters:
        //   value:
        //     An array of Unicode characters.
        public String(char[]? value);
        //
        // Summary:
        //     Initializes a new instance of the System.String class to the Unicode characters
        //     indicated in the specified read-only span.
        //
        // Parameters:
        //   value:
        //     A read-only span of Unicode characters.
        public String(ReadOnlySpan<char> value);

		...
    }
}
```

This would hopefully make reviewing API shapes much easier as you can just look at the diff in github (or your diff tool of choice). This would also have the added benefit of being able to see diffs of APIs across releases. At least if Roslyn's case, where we create a git tag for each release, you could diff two different tags to see the api changes. What are your thoughts @Evangelink ?

**Comment by jmarolf** [comment]
> The reason we didn't go with an analyzer to begin with was already described by @ericstj because we wanted binary-to-binary compatibility, however I do see this code being able to be expanded to take either a binary previously compiled or a `.cs` file as a contract and then running as an analyzer during a design time compilation. 

@safern This was the right decision for all the reasons you lay out: design-time feedback is likely niche compared to package validation in general. Would be curious to understand more specifically what scenarios a compilation-based approach would not catch vs. a binary approach. Today we have the api-compatibility analyzers and I would imagine a similar approach would have worked for package validation in some respect, though at the end of the day you really just care about what is going into a nuget package so I can see it being much simpler to start there and work backwards.

My current thinking is that we have two problems we are trying to solve
- breaking change detection
- api change review

I think if we can do the work to get package validation running on design-time builds that should meet our needs as a team for "breaking change detection." Breaking changes are something we plan to almost never do and this appears to fit well with the intended design of the package validation features. Would love to discuss what work we need to do on our side to feed into the package validation pipeline.

However, the whole "api change review" concept seems outside the scope of package validation and we may want to pivot the Public API analyzers to serve this purpose.

**Comment by Evangelink** [comment]
> What are your thoughts @Evangelink Amaury Levé FTE ?

That's awesome :) 

We need to be careful about the member ordering (I would expect the signature/api file to remain the same whatever the order in the src file) and maybe also about xml doc comments (some extensions are changing line breaks position for example).

> Breaking changes are something we plan to almost never do

I think that's a good decision. I think breaking change criteria are too specific. For example, if you really want to be strict adding an overload should be a breaking change as it could cause issues to F# (could be the case with capture and C#) but most people would not consider this as breaking change.

> * On CI you have some MSBuild property that gets passed to the analyzer/source generator to fail the build if the public API doesn't match the source (so if someone changes the `Public.Apis.csi` file and forgets to check it in we'll fail the build)

This would probably be a good fit for most users but it'd be nice to keep the door (or window) open for more advanced scenarios where you would want to make this an explicit test or make some specific changes to ignore parts or whatever advanced scenario people using `PublicApiGenerator` could have came up with (not saying we should handle these cases out-of-the-box nor in a first release).

**Comment by bitbonk** [comment]
For many years I have been involved in the maintenance of lot of libraries and frameworks are consumed in-house as well as by external customers, mainly via nuget packages. Here is my take on the essential user stories a good API tracking solution should cover:

1. As a library maintainer I want to be warned as early as possible (right in the IDE of my choice) when I have made a breaking change in the public API if that API already been shipped to customers.
2. As a library maintainer I want to get a good explanation what the breaking change is and why it is a breaking change, right in the IDE of my choice.
3. As a library maintainer I want to accept those breaking changes as deliberate and not be warned about those breaking changes from there on. 
4. As library maintainer who deliberately has to make a large amount of breaking changes, I want to have a way to quickly "bulk-accept" a large amount of warnings of a certain type right in the IDE of my choice.
5. As a library maintainer I want to quickly see in a PR whether the PR introduces (breaking and non-breaking) changes to the public API what those breaking changes are and whether it contains changes to APIs that have already been shipped to customers.
6. As a library maintainer I want to be able to create a CI build verification check that automatically fails when breaking changes  have been made.
7. As a library maintainer I want to mark new APIs as shipped to customer when I release my library, so I will be warned about breaking changes in the future.
8. As a library maintainer I want to provide (generate) documentation about the (breaking and non-breaking) changes in the public API for a new release of my library (aka [API Diff](https://github.com/dotnet/runtime/issues/62356)).
9. As a library consumer I want to determine what the breaking (breaking and non-breaking) changes the public API of a library has compared to the version I am currently using, even if the library maintainer did not provide an API diff documentation for it.

**Comment by bitbonk** [comment]
@jmarolf 
> * On CI you have some MSBuild property that gets passed to the analyzer/source generator to fail the build if the public API doesn't match the source (so if someone changes the `Public.Apis.csi` file and forgets to check it in we'll fail the build)

It is important that it still will be possible mark APIs as shipped at some point and that there is a way to make a build **not** fail, if I have only made changes to APIs that have not yet been shipped to customers. Will that still work?
